### ISSUE TITLE
Add forced coherency ability to "darc clone" rather than maintaining a dependency for every repo

### ISSUE THREAD

**Issue by dagood** [dotnet/source-build#1495]
To reasonably build source-build, we need a coherent set of repositories: one commit hash per repo. This rarely happens in the Microsoft build. When it does, it's mainly only for RTM, servicing, or preview builds, but it's not guaranteed even then.

We currently force coherency via a graph of CoherentParentDependency attributes in `eng/Version.Details.xml` in dotnet/source-build, where every repo has hand-crafted rules to pick the best version. We rely on updates from Maestro++ to apply these rules. `Version.Details.xml` is essentially a cache of how it's all resolved, allowing us to do a `darc clone --depth 0 ...` to fetch all the sources we need.

I chatted with @crummel about this--notes on an approach for the future:
* To unblock prototyping adding source-build directly into Core-SDK, I can simply copy the dependencies as they are into Core-SDK.
* We can preserve the coherency graph we've put together in some simplified form that we can give `darc clone`. It would use the coherency graph as guidance (hints) on how to resolve incoherency conflicts.
* Pushing this info into the repos is ideal. A repo should know how to determine which version of a dependency to use for source-build when it brings in two that are in conflict. This would make sense to have in `Version.Details.xml` in some way.
  * Currently, we would use BAR to look this up quickly without doing a *ton* of clones/checkouts to investigate the `Version.Details.xml` file in every single repo@commit. We don't have anonymous BAR access so we can't use this for local builds.
  * We can probably improve `darc clone` to grab `Version.Details.xml` without full checkouts to make this more feasible. (Definitely possible in Git, don't know about libgit.)

**Comment by adaggarwal** [comment]
Related #1470

**Comment by dagood** [comment]
I've been prototyping improvements to `darc clone` and seeing if we can actually get to a point where we can assemble the graph via info already available in the Version.Details.xml files. I think I'm at a point where I can conclude we should go ahead with some source-build-specific guiding info, rather than try to make something that "just works".

* We need to resolve the situation correctly when a repo appears in the graph multiple times with different commits.
  * In some cases "pick the latest asset version number" is trivially correct.
    * When a repo depends on `Microsoft.NETCore.App.Runtime.<rid>` 3.1.2 and older `Microsoft.NETCore.App.Ref` 3.1.0, pick 3.1.2, of course.
    * I believe this is the best we can do with an incoherent diamond dependency.
  * When we build stable, multiple commits exist for the same asset version number. Picking between those requires some sort of heuristic such as commit date.
    * If we disagree with the heuristic (e.g. an accidental stabilized build respin happens and we need to not use it) there needs to be some way to override the heuristic.
  * A system of heuristics that is complex *and* sometimes needs to be overridden is prone to human error. It seems it would cause us to use bad commits at some point and compromise the product vs. what Microsoft produced using prebuilts.
  * **Instead, I think we should build a system that is easy to understand, even if that system adds more source-build glue in repos.**
* We need some way to exclude dependencies on repos that aren't involved in source-build.
  * Typically, each repo excludes some PackageReferences and projects from the build, using MSBuild or build script logic. This means we can't reasonably detect which dependencies are used during source-build while running `darc clone`. (Certainly not detectable from Version.Details.xml alone.)
  * In source-build we have an ignore list that we pass into `darc clone` manually.
    * This ignore list should be moved to live close to, or ideally in, the repo with ignored dependencies. Not in a top-level orchestration repo.
  * **We need to make changes to repos' Version.Details.xml files to get this done the best way.**

Repo build order is something I haven't looked deeply at, but expect would not be reasonable to solve using magic either. Order depends on matching up asset dependencies with upstreams, which may be impossible when some dependencies are reference-only, or even completely unused but causing a circular dependency. **Adding info to Version.Details.xml seems right for this as well.**

Now I'm going into what we need to add to Version.Details.xml, and how we can start out with overrides in Core-SDK that migrate into the affected repos over time.

---

> * We can probably improve `darc clone` to grab `Version.Details.xml` without full checkouts to make this more feasible. (Definitely possible in Git, don't know about libgit.)

It is! DarcLib has an API that *should* support this that I was able to augment pretty easily to do what it says it should do. I've also added APIs for "bare" git clones and adding worktrees, to simplify and speed up source layout creation.

**Comment by dagood** [comment]
Right now I think marking specific dependencies with "ends up in the SDK product" is what we need. Then, during `darc clone`, the marked dependency's target repo overrides all dependencies on the same repo throughout the graph. If the same target repo is marked in two different dependencies in the graph as ending up in the SDK product, clone should fail.

Dependency update PRs may bring in a conflicting marking from upstream, so we need to make conflicts show up as a PR validation error, with a message that is actionable and easy to understand.

Override logic: we need a way for a repo to point deep into its upstream dependencies to add or remove the "ends up in the SDK product" marking. That way, we can go ahead with dotnet/installer without getting the new marking into all upstream repos first. Then, rollout can happen over time, phasing out the overrides.

For repos that don't end up in the SDK product, the "ends up in the SDK product" marking might not make sense anywhere. Non-shipping build tooling seems to fall into this category. (Arcade, sourcelink, and mono/linker might be good examples.) I think that in these cases, we should use the latest version that exists in the graph. This does mean we keep that extra complexity in our system, however it turns out this method is used for coherent parent dependencies, so I'm less concerned now about that being a big problem. We can mitigate the complexity by writing out graphviz files to visualize the resolver's decisions and reasoning. (I already started doing this for my own sake. 😄)

If there's a critical issue with resolving latest, and we must use an earlier version, the affected repo can mark the dependency on the earlier version as "ends up in the SDK product" to override.

I don't have a great name in mind for this marking yet. `ShipsInSdk` would be something for the main scenario. Something mechanic-driven (`SourceBuildCoherentizing`, `SourceBuildCritical`?) is applicable to emergency overrides too, but doesn't seem intuitive. Maybe `ProductCritical` combines both scenarios reasonably enough.

**Comment by dagood** [comment]
What I have so far seems good enough to me to get source-build into dotnet/installer (core-sdk), where we'd plop the `repos/*.proj`. I'm planning to submit a PR to arcade-services this week for this chunk of work.

I've split off [Determine repo build order during "darc clone" for arcade-powered source-build #1561](https://github.com/dotnet/source-build/issues/1561) so build order doesn't need to be manually maintained (e.g. included in the arcade SDK). That wouldn't be good for maintenance.

**Comment by dagood** [comment]
Had a meeting to discuss this with @mmitche and others, and thinking now we'll be able to rely on the CPD strictification of https://github.com/dotnet/arcade/issues/5195 and storing a complete coherent build graph in the dotnet/installer repo. That seems like it will remove the need for special darc clone features (until we want to do speculative builds).

The "rather than maintaining a dependency for every repo" in the title is kind of a non-goal here--as long as the dependencies don't require excessive maintenance, the extra list doesn't matter.

**Comment by dagood** [comment]
Closing since there's no further work planned in the original direction. What I've done so far is captured in this draft PR to arcade-services: https://github.com/dotnet/arcade-services/pull/1112, and investigatory stuff (e.g. graphs) in a powerpoint deck I shared in a meeting to explain the concept.

Opened https://github.com/dotnet/source-build/issues/1563 to track the new plan.
